// f5_nodejs/lib/ilx_lb_options.js
/*jshint node:true*/
'use strict';

/**
 * The set of options available for specifying the load balancing selection for a connection.
 * Note that for a load balancing selection to be made the plugin must enable the plugin
 * option ILXPluginOptions.handleClientOpen and not write any data to the server stream
 * ILXFlow.server until after the load balancing selection has been made.
 *
 * The options have the following precedence.
 *    - The system will load balance to the specified virtual server.
 *    - The system will load balance to the specified pool.
 *    - If neither a virtual server or pool are specified then the system will connect to
 *      the specified remote address.
 *    - If includeVirtualServers is enabled then a virtual server may be specified in the remote
 *      address settings.
 *
 * See also ILXFlow.lbSelection.
 *
 * @constructor
 *
 * @property {string} virtualServer - LTM virtual server name. The name must be a fully qualified
 *   path name, e.g., /Common/<vs name>.
 * @property {string} pool - LTM pool name. The name must be a fully qualified path name,
 *   e.g., /Common/<pool name>
 * @property {string} remote.address - Connect to the specified IP address. The address may
 *   contain a route domain, example: 10.1.1.1%22
 * @property {number} remote.port - Remote port.
 * @property {boolean} includeVirtualServers - Indicates that remote.address may be an LTM
 *   virtual server.
 *
 */
function ILXLbOptions () {

    this.virtualServer = "";
    this.pool = "";
    this.includeVirtualServers = false;

    this.remote = {
        address: "",
        port: 0,
    };
}

module.exports = ILXLbOptions;
